crypto/bls12381_utils.c

```diff
@@ -1,852 +1,1173 @@
+  }
+  E1 tmp;
+  E1_to_affine(&tmp, a);
+
+  Fp_from_montg(&tmp.x, &tmp.x);
+  Fp_write_bytes(out, &tmp.x);
+
+  if (G1_SERIALIZATION == COMPRESSED) {
+    out[0] |= (Fp_get_sign(&tmp.y) << 5);
+  } else {
+    Fp_from_montg(&tmp.y, &tmp.y);
+    Fp_write_bytes(out + Fp_BYTES, &tmp.y);
+  }
+  // compression bit
+  out[0] |= (G1_SERIALIZATION << 7);
 }
 
-// computes the sum of the G1 array elements y and writes the sum in jointy
-void ep_sum_vector(ep_t jointx, ep_st* x, const int len) {
-    ep_set_infty(jointx);
-    for (int i=0; i<len; i++){
-        ep_add_jacob(jointx, jointx, &x[i]);
-    }
+// generic point addition that must handle doubling and points at infinity
+void E1_add(E1 *res, const E1 *a, const E1 *b) {
+  POINTonE1_dadd((POINTonE1 *)res, (POINTonE1 *)a, (POINTonE1 *)b, NULL);
 }
 
-// Computes the sum of the signatures (G1 elements) flattened in a single sigs array
-// and writes the sum (G1 element) as bytes in dest.
-// The function assumes sigs is correctly allocated with regards to len.
-int ep_sum_vector_byte(byte* dest, const byte* sigs_bytes, const int len) {
-    int error = UNDEFINED;
-
-    // temp variables
-    ep_t acc;        
-    ep_new(acc);
-    ep_set_infty(acc);
-    ep_st* sigs = (ep_st*) malloc(len * sizeof(ep_st));
-    if (!sigs) goto mem_error;
-    for (int i=0; i < len; i++) ep_new(sigs[i]);
-
-    // import the points from the array
-    for (int i=0; i < len; i++) {
-        // deserialize each point from the input array
-        error = ep_read_bin_compact(&sigs[i], &sigs_bytes[SIGNATURE_LEN*i], SIGNATURE_LEN);
-        if (error != RLC_OK) {
-            goto out;
-        }
-    }
-    // sum the points
-    ep_sum_vector(acc, sigs, len);
-    // export the result
-    ep_write_bin_compact(dest, acc, SIGNATURE_LEN);
+// Point negation: res = -a
+void E1_neg(E1 *res, const E1 *a) {
+  E1_copy(res, a);
+  POINTonE1_cneg((POINTonE1 *)res, 1);
+}
 
-    error = VALID;
+// Exponentiation of a generic point `a` in E1, res = expo.a
+void E1_mult(E1 *res, const E1 *p, const Fr *expo) {
+  pow256 tmp;
+  pow256_from_Fr(tmp, expo);
+  POINTonE1_mult_glv((POINTonE1 *)res, (POINTonE1 *)p, tmp);
+  vec_zero(&tmp, sizeof(tmp));
+}
+
+// computes the sum of the E1 array elements `y[i]` and writes it in `sum`.
+void E1_sum_vector(E1 *sum, const E1 *y, const int len) {
+  E1_set_infty(sum);
+  for (int i = 0; i < len; i++) {
+    E1_add(sum, sum, &y[i]);
+  }
+}
+
+// Computes the sum of input E1 elements flattened in a single byte
+// array `in_bytes` of `in_len` bytes. and writes the sum (E1 element) as
+// bytes in `out`.
+// The function does not check membership of E1 inputs in G1
+// subgroup. The header is using byte pointers to minimize Cgo calls from the Go
+// layer.
+int E1_sum_vector_byte(byte *out, const byte *in_bytes, const int in_len) {
+  int error = UNDEFINED;
+  // sanity check that `len` is multiple of `G1_SER_BYTES`
+  if (in_len % G1_SER_BYTES) {
+    error = INVALID;
+    goto mem_error;
```

Comment:
seems like mem_error can be omitted and just return in both cases

Reply:
you're right, though I didn't want to mix the "goto" style with the "return early" style.